(GH-130) Add  documentation (#158)
puppetlabs/prm: docs/md/content/usage/validate.md

---
title: "Validating Puppet Content"
description: "Guidance for validating code with PRM."
category: narrative
tags:
- tools
- usage
- validation
weight: 10
---

In this guide you'll learn how the validate command works and how to use
it to quickly validate your Puppet content.

#### What is a validator
A validator is a tool which can read and validate a codebase of Puppet content.

The configuration of a validator is almost identical to that of a normal
tool. The only difference is that validators contain the `can_validate: true`
property in their `prm-config.yml` configuration file.

All validation tools in a tool directory can be listed with the following command:

```bash
prm validate --list
```

### How to use the validate command

#### Single tool validation

> It is very easy to run single tool validation with PRM. The command below shows an

Reviewer: ```suggestion
It is very easy to validate with a single tool. The command below shows an
```

> outline of the core args and flags necessary for content validation:

Reviewer: ```suggestion
outline of the core arguments and flags necessary for content validation:
```

```bash
prm validate {author/tool} --codedir {location/of/code} --toolArgs="{any, tool, args}"
```

> Example of a command composed to validate `epp` content against the current directory,

Reviewer: ```suggestion
Here is an example of a command composed to validate `epp` content against the current directory,
```

specifically, `dir1` and `dir2` as specified by the `toolArgs` flag.

```bash
prm validate puppetlabs/epp --codedir . --toolArgs="dir1, dir2"
```

When the command is executed PRM will validate with the `epp` tool using the appropriate tool

> args; the tool args in this case will instruct epp to validate content in the specified directories.

Reviewer: ```suggestion
arguments. The tool arguments in this case will instruct `epp` to validate content in the specified directories.
```

#### Multi-tool validation

PRM supports running multiple validation tools simultaneously. That's where the 
`validate.yml` file comes in; it allows you to arrange validators into groups.

> 

Reviewer: Maybe explain here why it can be useful to have a validate.yml file

Here's an example of a validate.yml file:

```yaml
groups:
  - id: "ci"
    tools:
      - name: puppetlabs/epp
        args: [--example, args]
      - name: puppetlabs/parser
      - name: puppetlabs/puppet-lint
      - name: puppetlabs/onceover
  - id: "quick_validate"
    tools:
      - name: puppetlabs/epp
      - name: puppetlabs/rubocop
      - name: puppetlabs/parser
      - name: puppetlabs/r10k
  - id: "syntax_validation"
    tools:
      - name: puppetlabs/puppet-syntax
      - name: puppetlabs/metadata-json-lint
      - name: puppetlabs/puppet-lint
```

> {{% alert title="Note" color="primary" %}}

Author: So the alerts will come through on documentation site looking like [this](https://www.docsy.dev/docs/adding-content/shortcodes/#alert); only the colour, heading and body will be different.

> {{% alert title="Note" color="primary" %}}

Author: The docs site can be run locally by running the `docs.sh` script in the root dir of PRM.

The `validate.yml` file should be stored in the root of the Puppet content codebase
which is to be validated.
{{% /alert %}}

> 

Reviewer: Could be nice to have an some sort of heading here

```bash
# No need to specify the --resultsView flag here as it defaults to `file` for mulit-tool validation
$ prm validate --codedir . --group syntax_validation
3:49PM INF Found tool group: syntax_validation 
3:49PM INF Validating with the metadata-json-lint tool
3:49PM INF Validating with the puppet-syntax tool
3:49PM INF Validating with the puppet-lint tool

      TOOL NAME      | VALIDATION EXIT CODE |                                 FILE LOCATION
---------------------+----------------------+--------------------------------------------------------------------------------
  puppet-syntax      |                    1 | .prm-validate/syntax_validation/puppet-syntax_2022_April_26_16-49-59.log
  metadata-json-lint |                    0 | .prm-validate/syntax_validation/metadata-json-lint_2022_April_26_16-49-59.log
  puppet-lint        |                    0 | .prm-validate/syntax_validation/puppet-lint_2022_April_26_16-49-59.log
3:49PM ERR Validation returned 1 error
```

{{% alert title="Note" color="primary" %}}
Support for formatting of the validation results will be implemented in a future 
release. E.g. JSON or JUNIT

> {{% /alert %}}

Reviewer 2: same again
